File: api/src/main/java/net/blueshell/api/repository/UserRepository.java
package net.blueshell.api.repository;

import net.blueshell.api.base.BaseRepository;
import net.blueshell.api.model.File;
import net.blueshell.api.model.User;
import org.springframework.data.jpa.repository.Query;
import org.springframework.stereotype.Repository;

import java.util.List;
import java.util.Optional;

@Repository
public interface UserRepository extends BaseRepository<User> {

    Optional<User> findByUsername(String username);

    boolean existsByUsername(String username);

    Optional<User> findByUsernameAndIdNot(String username, Long id);

    boolean existsByUsernameAndIdNot(String username, Long id);

    Optional<User> findByResetKey(String resetKey);

    boolean existsByResetKey(String resetKey);

    Optional<User> findByEmail(String email);

    boolean existsByEmail(String email);

    boolean existsByDiscord(String discord);

    boolean existsByDiscordAndIdNot(String discord, Long id);

    Optional<User> findByEmailAndIdNot(String email, Long id);

    boolean existsByEmailAndIdNot(String email, Long id);

    Optional<User> findByPhoneNumber(String phoneNumber);

    boolean existsByPhoneNumber(String phoneNumber);

    Optional<User> findByPhoneNumberAndIdNot(String phoneNumber, Long id);

    boolean existsByPhoneNumberAndIdNot(String phoneNumber, Long id);

    List<User> findByMembershipNotNull();

    Optional<User> findByMembershipSignature(File signature);

    Optional<User> findByProfilePicture(File profilePicture);
    @Query("SELECT DISTINCT u FROM User u " +
            "JOIN u.roles r " +
            "WHERE r = 'COMMITTEE'" +
            "AND NOT EXISTS (SELECT 1 FROM CommitteeMember cm WHERE cm.user = u)")
    List<User> findUsersWithRoleWithoutActiveCommittees();

    // Find users who do NOT have COMMITTEE role but ARE active committee members
    @Query("SELECT DISTINCT u FROM User u " +
            "WHERE NOT EXISTS (SELECT 1 FROM u.roles r WHERE r = 'COMMITTEE') " +
            "AND EXISTS (SELECT 1 FROM CommitteeMember cm WHERE cm.user = u)")
    List<User> findUsersWithoutRoleWithActiveCommittees();

    @Query("SELECT DISTINCT u FROM User u " +
            "JOIN u.roles r " +
            "WHERE r = 'MEMBER'" +
            "AND NOT EXISTS (SELECT 1 FROM Membership m WHERE m.user = u AND m.endDate IS NULL)")
    List<User> findUsersWithRoleWithoutActiveMembership();

    @Query("SELECT DISTINCT u FROM User u " +
            "WHERE NOT EXISTS (SELECT 1 FROM u.roles r WHERE r = 'MEMBER') " +
            "AND EXISTS (SELECT 1 FROM Membership m WHERE m.user = u AND m.endDate IS NULL)")
    List<User> findUsersWithoutRoleWithActiveMembership();
}

File: services/api/src/main/kotlin/net/blueshell/api/user/persistence/UserRepository.kt
package net.blueshell.api.user.persistence

import net.blueshell.api.shared.repository.BaseRepository
import org.springframework.data.domain.Pageable
import org.springframework.data.jpa.repository.Query
import org.springframework.data.repository.query.Param
import org.springframework.stereotype.Repository
import java.util.*

@Repository
interface UserRepository : BaseRepository<User, Long> {
    fun findByUsername(username: String): Optional<User>

    /** Ids of accounts that have not been activated. Soft-deleted rows are excluded by the entity. */
    @Query("select u.id from User u where u.enabled = false")
    fun findIdsByEnabledFalse(): List<Long>

    /** Everybody who opted into the newsletter, activated account or not. */
    @Query("SELECT u.id FROM User u WHERE u.newsletter = true")
    fun findIdsByNewsletterTrue(): List<Long>

    /**
     * `findAllById` with the profile fetched alongside. `User.memberProfile` is an eager
     * `mappedBy` one-to-one, so a lazily joined member costs an extra query each.
     */
    @Query(
        """
        SELECT u FROM User u
        LEFT JOIN FETCH u.memberProfile
        WHERE u.id IN :ids
        """,
    )
    fun findAllByIdsWithProfiles(@Param("ids") ids: Collection<Long>): List<User>

    fun existsByUsername(username: String): Boolean

    fun existsByUsernameAndIdNot(username: String, id: Long): Boolean

    fun existsByEmail(email: String): Boolean

    fun existsByDiscord(discord: String): Boolean

    fun existsByDiscordAndIdNot(discord: String, id: Long): Boolean

    fun existsByEmailAndIdNot(email: String, id: Long): Boolean

    fun existsByPhoneNumber(phoneNumber: String): Boolean

    fun existsByPhoneNumberAndIdNot(phoneNumber: String, id: Long): Boolean

    @Query(
        """
        SELECT CASE WHEN COUNT(m) > 0 THEN true ELSE false END
        FROM Membership m
        WHERE m.user.id = :userId
          AND m.endDate IS NULL
        """,
    )
    fun existsActiveMembershipByUserId(@Param("userId") userId: Long): Boolean

    /**
     * The id where the row exists at all, active or soft-deleted, so a caller can tell a user
     * who once existed from an id that never was. Native, to bypass `User`'s `@SQLRestriction`.
     */
    @Query(
        value = "SELECT id FROM users WHERE id = :userId AND deleted_at <> '9999-12-31 23:59:59.000000'",
        nativeQuery = true,
    )
    fun findSoftDeletedUserId(@Param("userId") userId: Long): Long?

    @Query(
        value = "SELECT id FROM users WHERE id IN (:userIds) AND deleted_at <> '9999-12-31 23:59:59.000000'",
        nativeQuery = true,
    )
    fun findSoftDeletedUserIds(@Param("userIds") userIds: Collection<Long>): List<Long>

    /**
     * Active user ids greater than [afterId], ascending — keyset pagination for
     * bulk fan-out that does not hold one transaction (or one big result set)
     * across the whole user table. Respects the `@SQLRestriction` on [User].
     */
    @Query("SELECT u.id FROM User u WHERE u.id > :afterId ORDER BY u.id")
    fun findActiveIdsAfter(@Param("afterId") afterId: Long, pageable: Pageable): List<Long>
}
